### ISSUE TITLE
container jobs fail with AssigningReplicaFailed error

### ISSUE THREAD

**Issue by jason-berk-k1x** [microsoft/azure-container-apps#1168]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is a: (mark with an x)
- [ ] bug report -> please search issues before submitting
- [x] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

### Issue description
Looking at my execution history, I see a ton of jobs that failed.  In an attempt to understand _why_ they failed, I ran this query on my logs:

```
ContainerAppSystemLogs_CL
| where TimeGenerated >= datetime(2024-01-01)
| where Type_s != 'Normal'
| where Reason_s != 'KEDAScalerFailed'
| summarize Count = count(), LatestOccurrence = max(TimeGenerated) by Reason_s
| project Reason_s, Count, LatestOccurrence
| order by Count desc
```

got back this result:

![Screenshot 2024-05-14 at 12 06 15 PM](https://github.com/microsoft/azure-container-apps/assets/148882427/95572afc-250b-49ec-9ed1-538069b8a321)

what is `AssigningReplicaFailed`?  I can't find any documentation as to why this is happening or what I can do to fix the issue.

### Steps to reproduce

setup a container app job that is triggered by KEDA when a queue has any messages

**Expected behavior** [What you expected to happen.]

some sort of documentation around these error Reasons

**Actual behavior** [What actually happened.]

I'm dead in the water with no idea why I'm getting this error so often

**Screenshots**         

![Screenshot 2024-05-14 at 12 12 00 PM](https://github.com/microsoft/azure-container-apps/assets/148882427/dd58a7f8-2179-4aa8-93a5-72699368196c)

![Screenshot 2024-05-14 at 12 14 06 PM](https://github.com/microsoft/azure-container-apps/assets/148882427/5840ff6d-855c-4749-851e-4b0ca77c21b4)

![Screenshot 2024-05-14 at 12 15 06 PM](https://github.com/microsoft/azure-container-apps/assets/148882427/ed6ad84d-c11a-4033-bc30-f7c3e3570069)

**Comment by jason-berk-k1x** [comment]
I opened a support ticket.... didn't here anything from anyone for over a week.  They finally got back to me on 5/24 and after a 90 minute phone call they weren't able to offer any help or insights as to what this error even means.  They said they were going to escalate my ticket 🤷

**Comment by jason-berk-k1x** [comment]
ran the query today just to have something to compare to:

![Screenshot 2024-06-17 at 5 26 06 PM](https://github.com/microsoft/azure-container-apps/assets/148882427/ee6daed7-9b39-41d2-bdae-16a19aa1139d)


so in 90 days I've gotten 25361 failures which is roughly 281 times a day

**Comment by pwarner** [comment]
We're having the same problem. Using the D16 workload profile, and we're seeing 50% of job executions fail.

**Comment by jason-berk-k1x** [comment]
my jobs are configured to timeout after 15 minutes and I poll the service bus queue ever 10 seconds.

my current theory is that we run into issues mostly when, for whatever reason, there's no capacity on the nodes that Azure manages and we have to wait for them to add more nodes to the pool....then pull the image...which is severely bloated at almost 9GiB, onto that new hardware....then start running the job.  Also, mind you, the 15 minute clock starts ticking when the message hits the queue and KEDA tries to start a job.... NOT when the job actually starts running

I've seen the *exact same processing* take three minutes when it runs on a node pool that already has pulled the image and has capacity to run a job.....then take 12 to 13 minutes when everything is starting cold

**Comment by jason-berk-k1x** [comment]
> You can avoid the issue by setting at least 1 node as minimum although minimum of 3 is recommended.

not sure I follow / agree with this.  I am running my jobs on the consumption profile.  I did experiment with setting the min to 1, instead of 0.  Because these are event triggered jobs, all that happens is a job is constantly running for 15 minutes (my timeout) and then dying because there's not any messages in the queue for the job to process.

> This does not happen on the consumption plan because there the jobs run on shared nodes, they are always running.

right.... many times I see my job start up and run "quickly" (under 2 minutes).  Occasionally it seems like the always running shared nodes don't have capacity for my job and that node pool scales up.....which takes time.....then the job gets scheduled....which takes time.....then the images get pulled....which takes more time (especially for my bloated images).....then the job finally starts.... and now has spent 10+ minutes of its life waiting and can only run for 5 more minutes before the timeout kills it

**Comment by jason-berk-k1x** [comment]
> This does not happen on the consumption plan because there the jobs run on shared nodes, they are always running.

right, those consumption plan nodes are always running.....but they might not have capacity to run your job...and now your job is just waiting for new nodes to be added to that node pool....or maybe for capacity to become available.  If new nodes are provisioned, then the job will likely need to pull the image fresh....adding more delay

**Comment by jason-berk-k1x** [comment]
as an aside, I'm also struggling with the math of the dedicated plans.  For example, if I create a new container app environment, and define a D4 profile, I can run my job on either:

- consumption plan with 4 cores / 8GB
- D4 profile with 4 cores / 16GB

on the consumption plan with max resources of 4 cores and 8GiB of RAM, at less than one cent / minute my 5 minute job will cost me $0.05

if I use the D4 profile instead......the price now depends on how you have the min/max nodes configured.  If I set the min to zero to emulate the consumption plan, then I don't feel like I'm really gaining anything because I still have to tolerate a cold node start if my jobs are intermittent and bursty (which they are).  Even after the code start, if I'm comparing apples to apples.....I'd be using the entire 4 cores of the D4 meaning I could only run one job per node.  I understand I could set my max nodes to 25 and totally have 25 nodes each running a single job...which is an experiment I'm about to go run.  

All that said, if I just keep one D4 node running, that would run me about $225 / month and the vast majority of that time, **NOTHING** would be running on that node.

I don't see how I could get anywhere near 4 cores and 8GB per job and have capacity for 25 jobs ([the app env has a soft limit of 100 cores](https://learn.microsoft.com/en-us/azure/container-apps/quotas#consumption-workload-profile)) outside of the consumption plan that's price comparable..... especially when

> The first 180,000 vCPU-seconds, 360,000 GiB-seconds, and 2 million requests each month are free. Beyond that, you pay for what you use on a per second basis based on the number of vCPU-s and GiB-s your applications are allocated.

**Comment by pwarner** [comment]
> the problem is, if you set Min Executions to 0 there is no node to schedule the container on. Upon start of the job the node will be initialized but that takes more time than the timeout. On the next run however it will work - until after some time the node is destroyed again and scaled down to 0.
> 
> You can avoid the issue by setting at least 1 node as minimum although minimum of 3 is recommended. This does not happen on the consumption plan because there the jobs run on shared nodes, they are always running.

Feels like this defeats the entire selling point of container apps jobs, that of being able to run (and scale) on demand and not pay for an instance to be running all the time. At the cost of keeping a D4 node running at all times, we might as well deploy azure durable functions on an app service plan, which is exactly what we have invested in getting away from.

**Comment by jason-berk-k1x** [comment]
@pwarner right..... there is no concept of "idle" job replicas.  The job starts and you pay full price while its running until it either completes or you hit your job timeout (which prevents runaway processes).  What I'm curious about is if you set the CPU resource settings for the job (not sure if those are "requests" or "limits") too high, if that actually causes issues with scheduling

for example, if I have job on the consumption plan I _can_ provision it 2 vcpu and 4 GiB of RAM.  When my job is triggered, the scheduler has to find a node with capacity to run that job.  If there is no node, I assume new nodes are added to the pool and then god only knows what sort of rebalancing happens to all the workloads across all the nodes.  Then images have to be pulled (assuming the job ends up on a node that didn't have the image cached) and all the while the job timeout countdown is ticking away.

If instead I provisioned only .25 vCores and 2 GiB, is the run time of my job impacted in any significant way?  Is my job more likely to be scheduled on existing nodes and avoid having to wait for hardware?

**Comment by pwarner** [comment]
> @pwarner right..... there is no concept of "idle" job replicas. The job starts and you pay full price while its running until it either completes or you hit your job timeout (which prevents runaway processes). What I'm curious about is if you set the CPU resource settings for the job (not sure if those are "requests" or "limits") too high, if that actually causes issues with scheduling
> 
> for example, if I have job on the consumption plan I _can_ provision it 2 vcpu and 4 GiB of RAM. When my job is triggered, the scheduler has to find a node with capacity to run that job. If there is no node, I assume new nodes are added to the pool and then god only knows what sort of rebalancing happens to all the workloads across all the nodes. Then images have to be pulled (assuming the job ends up on a node that didn't have the image cached) and all the while the job timeout countdown is ticking away.
> 
> If instead I provisioned only .25 vCores and 2 GiB, is the run time of my job impacted in any significant way? Is my job more likely to be scheduled on existing nodes and avoid having to wait for hardware?

@jason-berk-k1x 

I would love to know the answer to those questions as well. We raised a support ticket and asked them there too, but support just posted unhelpful links from documentation that doesn't answer provide any answers. 

@anthonychu kindly helped me out on twitter to understand what we had to do to play nice with the keda queue scalar, and I can see he added a label to this issue - hopefully he or someone else can expedite this.

**Comment by arthur-kerst** [comment]
@pwarner Could you share your insights from the discussion you had? We're also encountering this error, specifically for jobs that are executed and no nodes are available yet. We have a feeling that this is related to our jobs that get suspended unexpectedly.

We're using the D8 profile to schedule jobs with resources _cpu: 4 / memory 16Gi_ with minExecutions set to 0. If you can help us understand what we have to do to play nice with the KEDA queue scalar, that would be greatly appreciated!
